pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+				log.Warnf("DROP status returned from app while processing pub/sub event for entry id: %v", response.EntryID)
+				bulkSubDiag.statusWiseDiag[string(pubsub.Drop)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = nil
+			default:
+				// Consider unknown status field as error and retry
+				bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = errors.Errorf("unknown status returned from app while processing pub/sub event  for entry id %v: %v", response.EntryID, response.GetStatus())
+				hasAnyError = true
+			}
+		} else {
+			log.Warnf("Invalid entry id received from app while processing pub/sub event %v", response.EntryID)
+			continue
+		}
+	}
+	for _, item := range msg.entries {
+		ind := entryIDIndexMap[item.EntryID]
+		if (*bulkResponses)[ind].EntryID == "" {
+			(*bulkResponses)[ind].EntryID = item.EntryID
+			(*bulkResponses)[ind].Error = errors.Errorf("Response not received, RETRY required while processing bulk subscribe event for entry id: %v", item.EntryID)
+			hasAnyError = true
+			bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += 1
+		}
+	}
+	bulkSubDiag.elapsed = elapsed
+	if hasAnyError {
+		return errors.New("Few message(s) have failed during bulk subscribe operation")
+	} else {
+		return nil
+	}
+}
+
+// sendBulkToDeadLetter sends the bulk message to deadletter topic.
+func (a *DaprRuntime) sendBulkToDeadLetter(
+	ctx context.Context, name string, msg *pubsub.BulkMessage, deadLetterTopic string,
+	entryIDIndexMap *map[string]int, bulkResponses *[]pubsub.BulkSubscribeResponseEntry,
+	bulkSubDiag *bulkSubIngressDiagnostics,
+) error {
+	data := make([]pubsub.BulkMessageEntry, len(msg.Entries))
+
+	if bulkResponses == nil {
+		data = msg.Entries
+	} else {
+		n := 0
+		for _, message := range msg.Entries {
+			entryID := (*entryIDIndexMap)[message.EntryID]
+			if (*bulkResponses)[entryID].Error != nil {
+				data[n] = message
+				n++
+			}
+		}
+		data = data[:n]
+	}
+	bulkSubDiag.statusWiseDiag[string(pubsub.Drop)] += int64(len(data))
+	if bulkSubDiag.retryReported {
+		bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] -= int64(len(data))
+	}
+	req := &pubsub.BulkPublishRequest{
+		Entries:    data,
+		PubsubName: name,
+		Topic:      deadLetterTopic,
+		Metadata:   msg.Metadata,
+	}
+
+	_, err := a.BulkPublish(req)
+	if err != nil {
+		log.Errorf("error sending message to dead letter, origin topic: %s dead letter topic %s err: %w", msg.Topic, deadLetterTopic, err)
+	}
+	return err
+}
+
+func validateEntryID(entryID string, i int) error {
+	if entryID == "" {
+		log.Warn("Invalid blank entry id received while processing bulk pub/sub event, won't be able to process it")
+		return errors.New("Blank entryID supplied - won't be able to process it")
+	}
+	return nil
+}
+
+func populateBulkSubcribedMessage(message *pubsub.BulkMessageEntry, event interface{},
```

Comment:
Reviewers: Searches pubsubBulkSubscribedMessage for this path, and enters message in pubsubBulkSubscribedMessage to be sent to this path. If pubsubBulkSubscribedMessage is not found, creates one.